File: java/src/main/java/com/example/urlshortener/orchestrator/agents/TestingAgent.java
package com.example.urlshortener.orchestrator.agents;

import com.example.urlshortener.orchestrator.Node;
import com.example.urlshortener.orchestrator.SharedContext;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Path;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.TimeUnit;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Testing agent: actually shells out to Maven's test goal and reports a real
 * pass/fail result plus captured output -- Maven compiles before testing, so
 * this is also the compile gate for the whole pipeline (a distinctly-Java
 * failure mode Python doesn't have). Direct port of
 * orchestrator/agents/testing_agent.py.
 */
public final class TestingAgent {
    private TestingAgent() {}

    private static final Pattern SUMMARY_RE =
            Pattern.compile("Tests run: \\d+, Failures: \\d+, Errors: \\d+, Skipped: \\d+.*");

    public static Map<String, Object> runTests(SharedContext ctx, Node node, Path projectRoot) {
        String mvnExe = System.getProperty("os.name").toLowerCase().contains("win") ? "mvn.cmd" : "mvn";
        ProcessBuilder pb = new ProcessBuilder(mvnExe, "-o", "test");
        pb.directory(projectRoot.toFile());
        pb.redirectErrorStream(true);

        String output;
        int exitCode;
        try {
            Process proc = pb.start();
            output = new String(proc.getInputStream().readAllBytes());
            boolean finished = proc.waitFor(180, TimeUnit.SECONDS);
            if (!finished) {
                proc.destroyForcibly();
                throw new RuntimeException("mvn test timed out after 180s");
            }
            exitCode = proc.exitValue();
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new RuntimeException(e);
        }

        boolean passed = exitCode == 0;
        Matcher m = SUMMARY_RE.matcher(output);
        String summaryLine = null;
        while (m.find()) summaryLine = m.group(); // keep the last (aggregate) match
        if (summaryLine == null) {
            summaryLine = passed ? "BUILD SUCCESS" : "BUILD FAILURE (see stdout_tail for details)";
        }

        ctx.recordDecision(
                node.stage, node.id,
                "mvn test " + (passed ? "passed" : "failed") + ": " + summaryLine,
                "Ran the real build+test as an exit gate (Maven compiles before testing, so this also catches "
                        + "compile errors); downstream docs/release nodes are policy-blocked from proceeding on a "
                        + "failing result."
        );

        Map<String, Object> out = new LinkedHashMap<>();
        out.put("passed", passed);
        out.put("exit_code", exitCode);
        out.put("summary", summaryLine);
        out.put("stdout_tail", output.length() > 4000 ? output.substring(output.length() - 4000) : output);
        return out;
    }
}

File: orchestrator/agents/testing_agent.py
"""Testing agent: actually shells out to pytest and reports a real pass/fail
result plus captured output -- this is a genuine test-execution gate, not a
simulated one. The `tests_must_pass` exit policy (orchestrator/policy.py) is
what turns a failing result into a blocked node.
"""
import subprocess
import sys


def run_tests(ctx, node, test_path: str, project_root: str, python_exe: str | None = None) -> dict:
    exe = python_exe or sys.executable
    proc = subprocess.run(
        [exe, "-m", "pytest", test_path, "-q"],
        cwd=project_root,
        capture_output=True,
        text=True,
        timeout=120,
    )
    passed = proc.returncode == 0
    summary_line = next(
        (line for line in reversed(proc.stdout.splitlines()) if line.strip()), proc.stdout[-200:] if proc.stdout else ""
    )

    ctx.record_decision(
        stage=node.stage,
        node_id=node.id,
        summary=f"pytest {test_path} {'passed' if passed else 'failed'}: {summary_line}",
        rationale="Ran the real test suite as an exit gate; downstream docs/release nodes are policy-blocked "
        "from proceeding on a failing result.",
    )

    return {
        "passed": passed,
        "returncode": proc.returncode,
        "summary": summary_line,
        "stdout_tail": proc.stdout[-3000:],
        "stderr_tail": proc.stderr[-1500:],
    }
